### PR TITLE
CA2243 attribute string literal should parse correctly

### PR DESCRIPTION
**Pull request by basoundr**
Implement CA2243 AttributeLiterals should Parse Correctly

If the Attribute Constructor is taking a Guid or Uri then verify the literals
can actually be parsed into Guids and Uris

Tagging @dotnet/roslyn-analysis for review

### CODE REVIEW

File: src/Test/Utilities/DiagnosticAnalyzerTestBase.cs [hunk 1]
```diff
@@ -87,7 +87,7 @@ public static MetadataReference SystemThreadingTaskFacadeRef
             }
         }
 
-        protected bool PrintActualDiagnosticsOnFailure { private get; set; }
+        protected bool PrintActualDiagnosticsOnFailure { get; set; }
```
Comment (srivatsn) [inline src/Test/Utilities/DiagnosticAnalyzerTestBase.cs]: Nice. CA1044 triggered this change? :)

File: src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
-using Analyzer.Utilities;
+using Microsoft.CodeAnalysis.Semantics;
 
 namespace System.Runtime.Analyzers
 {
     /// <summary>
     /// CA2243: Attribute string literals should parse correctly
     /// </summary>
```
Comment (srivatsn) [inline src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs]: Add a comment here explaining that this rule doesn't warn on version like FxCop used to and the reason for the deviation.

File: src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs [hunk 3]
```diff
@@ -25,24 +29,232 @@ public abstract class AttributeStringLiteralsShouldParseCorrectlyAnalyzer : Diag
...
+                            {
+                                AnalyzeSymbols(saContext.ReportDiagnostic, methodSymbol.Parameters);
+                                AnalyzeSymbols(saContext.ReportDiagnostic, methodSymbol.TypeParameters);
+                            }
+
+                            var propertySymbol = symbol as IPropertySymbol;
+                            if (propertySymbol != null)
+                            {
+                                AnalyzeSymbols(saContext.ReportDiagnostic, propertySymbol.Parameters);
+                            }
```
Comment (srivatsn) [inline src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs]: Curious - why are these not in their own case statements?

**Comment by srivatsn** [comment]
:+1: modulo a couple of suggestions.

File: src/Test/Utilities/DiagnosticAnalyzerTestBase.cs (hunk 1 above)
Comment (basoundr) [inline src/Test/Utilities/DiagnosticAnalyzerTestBase.cs]: Conversation about this rule made me make this change

File: src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs (hunk 2 above)
Comment (basoundr) [inline src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs]: Sure

File: src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs (hunk 3 above)
Comment (basoundr) [inline src/System.Runtime.Analyzers/Core/AttributeStringLiteralsShouldParseCorrectly.cs]: They can be. As I was addressing different case, the code diverged. I will modify the code.

**Pull request merged**